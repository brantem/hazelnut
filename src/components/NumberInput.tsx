import { MinusCircleIcon, PlusCircleIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';

type NumberInputProps = {
  label?: string;
  name: string;
  value: number;
  onChange: (quantity: number) => void;
  min?: number;
  max?: number;
  step?: number;
  isDisabled?: boolean;
};

const NumberInput = ({ label, name, value = 0, onChange, min = 0, max, step = 1, isDisabled }: NumberInputProps) => {
  return (
    <div className="flex w-full items-center justify-between space-x-3">
      {label && (
        <label htmlFor={name} className="flex-1 truncate font-medium">
          {label}
        </label>
      )}

      <div className="flex items-center space-x-1">
        <button
          type="button"
          className="rounded-full disabled:opacity-50"
          onClick={() => onChange(value - step)}
          disabled={isDisabled || value === min}
          data-testid="number-input-decrement"
        >
          <MinusCircleIcon className="h-6 w-6" />
        </button>

        <input
          type="number"
          className={clsx(
            'border-0 bg-transparent p-0 text-center text-sm tabular-nums disabled:text-neutral-500',
            /* c8 ignore next */ value.toString().length > step ? 'w-7' : 'w-5',
          )}
          name={name}
          id={name}
          value={value}
          onChange={(e) => onChange?.(parseInt(e.target.value))}
          min={min}
          max={max}
          step={step}
          disabled={isDisabled}
          data-testid="number-input-value"
        />

        <button
          type="button"
          className="rounded-full disabled:opacity-50"
          onClick={() => onChange(value + step)}
          disabled={isDisabled || Boolean(max && value >= max)}
          data-testid="number-input-increment"
        >
          <PlusCircleIcon className="h-6 w-6" />
        </button>
      </div>
    </div>
  );
};

export default NumberInput;
