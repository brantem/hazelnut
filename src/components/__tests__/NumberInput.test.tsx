import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';

import NumberInput from 'components/NumberInput';

test('NumberInput', () => {
  const onChange = vi.fn(() => {});
  const props = { label: 'Label', name: 'name', min: 0, max: 2, step: 2, value: 0, onChange };
  const { container, rerender } = render(<NumberInput {...props} />);

  expect(container).toMatchSnapshot();
  expect(screen.getByTestId('number-input-decrement')).toBeDisabled();
  act(() => screen.getByTestId('number-input-increment').click());
  expect(onChange).toHaveBeenCalledWith(2);
  rerender(<NumberInput {...props} value={2} />);
  expect(screen.getByTestId('number-input-increment')).toBeDisabled();
  act(() => screen.getByTestId('number-input-decrement').click());
  expect(onChange).toHaveBeenCalledWith(0);
});
