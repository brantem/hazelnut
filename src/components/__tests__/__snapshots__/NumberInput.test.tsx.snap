// Vitest Snapshot v1

exports[`NumberInput 1`] = `
<div>
  <div
    class="flex w-full items-center justify-between space-x-3"
  >
    <label
      class="flex-1 truncate font-medium"
    >
      Label
    </label>
    <div
      class="!-mr-0.5 flex items-center space-x-1"
    >
      <button
        class="rounded-full disabled:opacity-50 text-mono-600"
        data-testid="number-input-decrement"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15 12H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <span
        class="px-1.5 text-center text-sm tabular-nums text-mono-600 rounded-full"
        data-testid="number-input-value"
      >
        0
      </span>
      <button
        class="rounded-full disabled:opacity-50 text-mono-600"
        data-testid="number-input-increment"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 9v6m3-3H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
